google-cloud-storage/src/main/java/com/google/cloud/storage/UnifiedOpts.java

```diff
@@ -0,0 +1,1438 @@
+      super(StorageRpc.Option.IF_SOURCE_GENERATION_MATCH, val);
+    }
+
+    @Override
+    public Mapper<GetObjectRequest.Builder> getObject() {
+      return Mapper.identity();
+    }
+  }
+
+  /**
+   * @see GenerationNotMatch
+   * @see SourceGenerationMatch
+   */
+  static final class SourceGenerationNotMatch extends RpcOptVal<@NonNull Long>
+      implements ObjectSourceOpt, ObjectTargetOpt {
+    private static final long serialVersionUID = 313414895558156715L;
+
+    private SourceGenerationNotMatch(@NonNull Long val) {
+      super(StorageRpc.Option.IF_SOURCE_GENERATION_NOT_MATCH, val);
+    }
+
+    @Override
+    public Mapper<GetObjectRequest.Builder> getObject() {
+      return Mapper.identity();
+    }
+  }
+
+  /**
+   * @see MetagenerationMatch
+   * @see SourceMetagenerationNotMatch
+   */
+  static final class SourceMetagenerationMatch extends RpcOptVal<@NonNull Long>
+      implements BucketSourceOpt, BucketTargetOpt, ObjectSourceOpt, ObjectTargetOpt {
+    private static final long serialVersionUID = -3643340315457580094L;
+
+    private SourceMetagenerationMatch(@NonNull Long val) {
+      super(StorageRpc.Option.IF_SOURCE_METAGENERATION_MATCH, val);
+    }
+
+    @Override
+    public Mapper<GetObjectRequest.Builder> getObject() {
+      return Mapper.identity();
+    }
+  }
+
+  /**
+   * @see MetagenerationNotMatch
+   * @see SourceMetagenerationMatch
+   */
+  static final class SourceMetagenerationNotMatch extends RpcOptVal<@NonNull Long>
+      implements BucketSourceOpt, BucketTargetOpt, ObjectSourceOpt, ObjectTargetOpt {
+    private static final long serialVersionUID = -6682202521743160969L;
+
+    private SourceMetagenerationNotMatch(@NonNull Long val) {
+      super(StorageRpc.Option.IF_SOURCE_METAGENERATION_NOT_MATCH, val);
+    }
+
+    @Override
+    public Mapper<GetObjectRequest.Builder> getObject() {
+      return Mapper.identity();
+    }
+  }
+
+  static final class RequestedPolicyVersion extends RpcOptVal<@NonNull Long>
+      implements BucketSourceOpt {
+    private static final long serialVersionUID = 7044856817626952830L;
+
+    private RequestedPolicyVersion(Long val) {
+      super(StorageRpc.Option.REQUESTED_POLICY_VERSION, val);
+    }
+  }
+
+  static final class ReturnRawInputStream extends RpcOptVal<@NonNull Boolean>
+      implements ObjectSourceOpt {
+    private static final long serialVersionUID = 505293506385742781L;
+
+    private ReturnRawInputStream(boolean val) {
+      super(StorageRpc.Option.RETURN_RAW_INPUT_STREAM, val);
+    }
+
+    @Override
+    public Mapper<GetObjectRequest.Builder> getObject() {
+      return todo();
+    }
+  }
+
+  static final class ServiceAccount extends RpcOptVal<String> implements HmacKeyListOpt {
+    private static final long serialVersionUID = 1630581690347694016L;
+
+    private ServiceAccount(String val) {
+      super(StorageRpc.Option.SERVICE_ACCOUNT_EMAIL, val);
+    }
+  }
+
+  static final class SetContentType implements ObjectTargetOpt {
```

Comment:
We have a lot of boiler plate in these classes; would it be reasonable to introduce Autoclass?

Reply:
Using AutoValue would prevent us from defining final classes which I really don't want to do on a (likely future) public api.

The majority of our `Opt`s extend off `RpcOptVal` so they only need a constructor, it's mainly these more esoteric extractor/mutation `Opt`s which are needing their own `equals` and `hashCode`.